diff --git a/src/libexpr/flake/flake.cc b/src/libexpr/flake/flake.cc
index 119c556acdf..f55a98ac21b 100644
--- a/src/libexpr/flake/flake.cc
+++ b/src/libexpr/flake/flake.cc
@@ -175,6 +175,7 @@ static std::map<FlakeId, FlakeInput> parseFlakeInputs(
 {
     std::map<FlakeId, FlakeInput> inputs;
 
+    state.forceValue(*value, pos);
     expectType(state, nAttrs, *value, pos);
 
     for (nix::Attr & inputAttr : *(*value).attrs) {
